File: md.cpp
//#include <iostream>
#include <iomanip>
#include <cstdlib>
#include <fstream>
#include <stdlib.h>
#include <math.h>
#include <cmath>
#include <time.h>
#include "gasdev.cpp"
#include "other/nr.h"
#define  LengthX 3.18198052
#define  LengthY 3.18198052
#define  LengthZ 4.5
//#define  displacement  1.5
#define  Number 54    //number of atoms should be   N=2*(n)^3 ==>> "n" should be integer
#define  Cutoff 5   //cut off
#define  dump_every 100
#define  log_every  100  
#define  thermostat_every  100000000
#define  h          0.0005
#define  desired_Temperature          20 //K
#define  kB_True         1.38064852e-23  //m2 kg s-2 K-1
#define  epsilon_True    1.666e-21 //J
#define  sigma_True      3.4e−10    //m
/* MD units 
  kB         1 
  epsilon    1
  sigma      1
 
 * */
using namespace std;
void Force(float [][3],float [][3]);
float KEnergy(float [][3]);
float PEnergy(float [][3]);
void velocity_verlet(float [][3],float [][3],float [][3], int);
void dump(float [][3],int, ofstream& );
void log(float[][3],float[][3],int, ofstream&);
void pbc(float[][3], int);
float Temperature(float [][3]);
void Thermostat_velocity_scaling(float [][3]);
int main(){
    
    float X[Number][3],V[Number][3],F[Number][3];
    float r,T,t,m,YMassVelocity=0,XMassVelocity=0 ,ZMassVelocity=0;
    float X_cm=0, Y_cm=0, Z_cm=0;
    int i,j,k,N,a,step=0,idum=-8; 
    ofstream M_output_file ("dump.xyz");
    ofstream E_output_file ("log.dat");
    cout<<"inter full time:\n ";
//     cin>>T;
    T = 1000000;

// Old positioning block!!!!!!!!!

//    k=sqrt(Number/2)+1;
////     m=(Length/2)/(sqrt(Number/2)+1);
//    m= displacement;
//    N=Number;
//        for(j=0;j<=Number/k;j++){
//            for(i=1;i<=k;i++){
//                if (N!=0){
//                X[Number-N][0]=i*m;
//                X[Number-N][1]=(j+1)*m;
//                X_cm += X[Number-N][0]/Number;
//                Y_cm += X[Number-N][1]/Number;
//                N=N-1;
//                }
//            }
//    }

//new positioning
    system( (" python create_argon.py -n '"+to_string(int(pow(Number/2,1./3)))+" '").c_str() ); // '"+name+"  '
    ifstream myfile ("initial_positions_"+to_string(Number)+"_atoms.xyz");
    float boxsize_X, boxsize_Y, boxsize_Z;
    myfile >> boxsize_X;
    myfile >> boxsize_Y;
    myfile >> boxsize_Z;
//    cout<<"yes!\t"<<boxsize_X<<"\t"<<boxsize_Y<<"\t"<<boxsize_Z<<"\n";
    for(j=0;j<Number;j++){
        myfile >> X[j][0];
        myfile >> X[j][1];
        myfile >> X[j][2];
        X_cm += X[j][0]/Number;
        Y_cm += X[j][1]/Number;
        Z_cm += X[j][2]/Number;
        }
    myfile.close();
//


//     cout<<X_cm<<"\t"<<Y_cm<<endl;
//     // move center of mass to the middle:
//    for(i=0;i<Number;i++){
//        X[i][0] += (LengthX/2-X_cm);
//        X[i][1] += (LengthY/2-Y_cm);
//        X[i][1] += (LengthZ/2-Z_cm);
//    }

    dump(X,step,M_output_file);
    //end of first place calculation...
    
    for(i=0;i<Number;i++){
        V[i][0]=NR::gasdev(idum);
        V[i][1]=NR::gasdev(idum);
        V[i][2]=NR::gasdev(idum);
    }
	for(i=0;i<Number;i++){                       //calculate Central Mass Velocity
		XMassVelocity+=V[i][0];
		YMassVelocity+=V[i][1];
		ZMassVelocity+=V[i][2];
	}
	for(i=0;i<Number;i++){                       //Central Mass Velocity=0
		V[i][2]=V[i][2]-ZMassVelocity/Number;
		V[i][1]=V[i][1]-YMassVelocity/Number;
		V[i][0]=V[i][0]-XMassVelocity/Number;
	}
	
	Thermostat_velocity_scaling(V);
	
	
	Force(X,F); // calculate a0
	
        /// MAIN ///
        clock_t tStart = clock();
        for(step=1;step<T;step+=1){
            velocity_verlet(V,X,F, step);
            dump(X,step,M_output_file);
            log(X,V,step,E_output_file);
            if(step%thermostat_every==0)        Thermostat_velocity_scaling(V);
        }
        
    
        printf("Time taken: %.2fs\n", (double)(clock() - tStart)/CLOCKS_PER_SEC);
        
return 0;
}

void pbc(float X[][3], int i){
    X[i][0] -= LengthX*floor(float(X[i][0]/LengthX));
    X[i][1] -= LengthY*floor(float(X[i][1]/LengthY));
    X[i][2] -= LengthZ*floor(float(X[i][2]/LengthZ));

//     if(X[i][0]>Length)  X[i][0]+=(-1)*Length;    //boundary conditions...
//     if(X[i][0]<0)       X[i][0]+=Length;		 	//boundary conditions...	
//     if(X[i][1]>Length)  X[i][1]+=(-1)*Length;    //boundary conditions...
//     if(X[i][1]<0)       X[i][1]+=Length;         //boundary conditions...
}


void Force(float X[][3],float F[][3]){             //calculate acceleration
    int i,j;
    float r2,deltax,deltay, deltaz, solid_distanse_x,solid_distanse_y ,solid_distanse_z;
    float LengthX_2 = float(LengthX/2);
    float LengthY_2 = float(LengthY/2);
    float LengthZ_2 = float(LengthZ/2);

    for(i=0;i<Number;i++){
        F[i][0]=0;
        F[i][1]=0;
        F[i][2]=0;
//        cout<<X[i][0]<<X[i][1]<<"\n";
    }
//    cout<<"\n\n\n\n";
	for(i=0;i<Number;i++){
	    for(j=0;j<Number;j++){
                if(j!=i){ //this IF is unnecessary instead set i>j in the loop
                    
                    solid_distanse_x = abs(X[i][0]-X[j][0]);
                    deltax = -solid_distanse_x + LengthX*floor(float(solid_distanse_x/LengthX_2));
                                
                    solid_distanse_y = abs(X[i][1]-X[j][1]);
                    deltay=  -solid_distanse_y + LengthY*floor(float(solid_distanse_y/LengthX_2));

                    solid_distanse_z = abs(X[i][2]-X[j][2]);
                    deltaz=  -solid_distanse_z + LengthZ*floor(float(solid_distanse_z/LengthX_2));

                    r2=pow(deltax,2)+ pow(deltay,2) +pow(deltaz,2);
                    
                    
                    if(sqrt(r2)<Cutoff){
                        if (X[i][0]<X[j][0])     F[i][0]+=(deltax*48*(1/pow(r2,7)-1/(2*pow(r2,4))));   //  note that  r2=r*r
                        else                     F[i][0]+=(-1*deltax*48*(1/pow(r2,7)-1/(2*pow(r2,4))));

                        if (X[i][1]<X[j][1])     F[i][1]+=(deltay*48*(1/pow(r2,7)-1/(2*pow(r2,4))));
                        else                     F[i][1]+=(-1*deltay*48*(1/pow(r2,7)-1/(2*pow(r2,4))));

                        if (X[i][2]<X[j][2])     F[i][2]+=(deltaz*48*(1/pow(r2,7)-1/(2*pow(r2,4))));
                        else                     F[i][2]+=(-1*deltaz*48*(1/pow(r2,7)-1/(2*pow(r2,4))));
                    }
                }
            }
    }
}



// void Force(float X[][3],float F[][3]){             //calculate acceleration
//     int i,j,marzx,marzy;
//     float r2,deltax,deltay;
//     for(i=0;i<Number;i++){
// 		F[i][0]=0;
// 		F[i][1]=0;
//     }
// 	for(i=0;i<Number;i++){
// 	    for(j=0;j<Number;j++){
//     	    if(j!=i){
//         	    if(abs(X[i][0]-X[j][0])>Length/2){
//             	  deltax=Length-abs(X[i][0]-X[j][0]);
// 				  marzx=0;
// 				}
//             	else{
//                 	deltax=abs(X[i][0]-X[j][0]);
//                 	marzx=1;
// 				}
//             	if(abs(X[i][1]-X[j][1])>Length/2){
//             		deltay=Length-abs(X[i][1]-X[j][1]);
//               		marzy=0;
// 				}
//               	else{
//                 	deltay=abs(X[i][1]-X[j][1]);
//                 	marzy=1;
// 				}
//         		r2=pow(deltax,2)+pow(deltay,2);
//         		if(sqrt(r2)<Cutoff){
//                 	if (X[i][0]<X[j][0]) F[i][0]+=(pow(-1,marzx)*deltax*48*(1/pow(r2,7)-1/(2*pow(r2,4))));   //  note that  r2=r*r
//                 	else F[i][0]+=(pow(-1,marzx+1)*deltax*48*(1/pow(r2,7)-1/(2*pow(r2,4))));
//                 	if (X[i][1]<X[j][1])  F[i][1]+=(pow(-1,marzy)*deltay*48*(1/pow(r2,7)-1/(2*pow(r2,4))));
//                 	else F[i][1]+=(pow(-1,marzy+1)*deltay*48*(1/pow(r2,7)-1/(2*pow(r2,4))));
// 				}
// 			}
// 		}
//                
// 	}
// }

void Thermostat_velocity_scaling(float V[][3]){
    float lambda;
    int i;
    float temperature_True = float(epsilon_True)/kB_True;  // multiply temperature with this value to take in Kelvin

    float temperature_now =Temperature(V)*temperature_True;
//     cout<<"temperature_now="<<temperature_now<<endl;
    lambda = sqrt(desired_Temperature/temperature_now);
    for(i=0;i<Number;i++){
        V[i][0] *= lambda;
        V[i][1] *= lambda;
        V[i][2] *= lambda;
    }
    }

void velocity_verlet(float V[][3], float X[][3], float F[][3], int step){
    	                                       
    // verlet loop //
    int i;
    for(i=0;i<Number;i++){                             //calculate v half step
        V[i][0]+=(h/2)*(F[i][0]);
        V[i][1]+=(h/2)*(F[i][1]);
        V[i][2]+=(h/2)*(F[i][2]);
        X[i][0]+=V[i][0]*h;
        X[i][1]+=V[i][1]*h;
        X[i][2]+=V[i][2]*h;
        pbc(X,i);
    }
    Force(X,F);                                         //calculate a
    for(i=0;i<Number;i++){                               //calculate v half step
        V[i][0]+=(h/2)*(F[i][0]);
        V[i][1]+=(h/2)*(F[i][1]);
        V[i][2]+=(h/2)*(F[i][2]);
    }
        //cout<<step<<endl;
    }

void log(float X[][3],float V[][3],int step, ofstream& E_output_file){
    //cout<<step%log_every<<endl;
    if(step%log_every==0){
        float temrature_True = float(epsilon_True)/kB_True;  // multiply temperature with this value to take in Kelvin
        float kinetic_E   = KEnergy(V)*epsilon_True;       //calculate kinetic Energy
        float potential_E = PEnergy(X)*epsilon_True;    	//calculate potential Energy
        float temperature_now = Temperature(V)*temrature_True;
        E_output_file<<step<<"\t"<<kinetic_E<<"\t"<<potential_E<<"\t"<<kinetic_E+potential_E<<"\t"<<temperature_now<<endl;
        //cout<<"loged at step="<<step<<endl;
    }
    }

void dump(float X[][3], int step, ofstream& M_output_file){
    if(step%dump_every==0){
        M_output_file<<Number<<endl<<"atoms"<<endl; 
        for(int i=0;i<Number;i++){
            M_output_file<<1<<"\t"<<X[i][0]<<"\t"<<X[i][1]<<"\t"<<X[i][2]<<endl;
        }
//         cout<<"dumped at step="<<step<<endl;
    }
    }

float Temperature(float V[][3]){
    return KEnergy(V) * float(1)/(float(3)/2*Number);
    }

float KEnergy(float V[][3]){           //calculate kinetic Energy
    int i;
    float E=0;
    for(i=0;i<Number;i++){
    	E+=.5*(pow(V[i][0],2)+pow(V[i][1],2) +pow(V[i][2],2) );
    }
    return E;
    }
    
float PEnergy(float X[][3]){             //calculate potential Energy
    float r,deltax,deltay, deltaz,E=0;
    int j,i;
    for(i=0;i<Number;i++){
        for(j=0;j<Number;j++){
             if(j!=i){ // this IF is unnecessary use j=i instead
             	if(abs(X[i][0]-X[j][0])>LengthX/2){
            	  deltax=LengthX-abs(X[i][0]-X[j][0]);
				}
            	else{
                	deltax=abs(X[i][0]-X[j][0]);
				}


            	if(abs(X[i][1]-X[j][1])>LengthY/2){
            		deltay=LengthY-abs(X[i][1]-X[j][1]);
				}
              	else{
                	deltay=abs(X[i][1]-X[j][1]);
				}

				if(abs(X[i][2]-X[j][2])>LengthZ/2){
            		deltaz=LengthZ-abs(X[i][2]-X[j][2]);
				}
              	else{
                	deltaz=abs(X[i][2]-X[j][2]);
				}

        		r=sqrt(pow(deltax,2)+pow(deltay,2) +pow(deltaz,2)  );
                if(r<Cutoff){
                E+=4*(1/pow(r,12)-1/pow(r,6));
                }
		    }
		}
	}
    E=0.5*E;        // because of Uij=Uji
    
    return E;
    }

    

File: md.py
#!/usr/bin/env python3
import numpy as np
import time
import os
from numba import njit,prange
from ase import visualize
from ase.build import stack
from ase.io import write
import numpy as np
from ase.build import bulk
np.random.seed(4) # debug


# hardcoded until numba improves support for classes
def initial_parameteres():
    global Natoms,X0, n_a,n_b,n_c,box_sizes,box_half_sizes,Nsteps,cutoff,dump_step,log_step,velocity_zeroing_step,temp_ref,temp_step,Potential_formula,kB_true
    global trajectory_file,trajectory_file_unwrapped, log_file

    Natoms = 0 # will be set, also as global value in premain
    X0=0
    n_a = 3  #Natoms of primitive cell in a direction
    n_b = 3  #Natoms of primitive cell in b direction
    n_c = 3  #Natoms of primitive cell in c direction
    box_sizes = [0,0,0] # from 0 to L in each direction. will be set, also as global value in premain
    box_half_sizes = 0 # will be set, also as global value in premain
    Nsteps = 10**5
    cutoff = 3   #cut off
    dump_step = 100
    log_step = 10
    velocity_zeroing_step =100
    # dt = 0.0005
    temp_ref = 20 # reference tempreature in Kelvin
    temp_step = 100 # thermostat every N steps

    Potential_formula= 'LJ' # Potential type 'LJ' or 'Morse'

    kB_true = 1.38064852e-23  #m2 kg s-2 K-1

    trajectory_file = "traj.xyz"
    trajectory_file_unwrapped = "traj_unwrapped.xyz"
    log_file = "output.dat"

    # constants
    # kBTrue = 1.38064852e-23  #m2 kg s-2 K-1
    # epsilon_True = 1.65e-21 #J
    # sigma_True = 3.4e-10    #m

     
    global force,potential_energy
    global D0,r0,alpha,electronvolt_to_joules,epsilon_true,sigma_true
    global mass,tau,dt
    if Potential_formula == 'Morse':
        D0 =  0.3429 # ev  ## D in Morse potential ## FOR Copper ##
        r0 =  2.866e-10 # meter
        alpha = 1.3588e10 # 1/meter
        # map to LJ units in order to avoid changing the code :)
        electronvolt_to_joules = 1.60218e-19
        epsilon_true = D0 * electronvolt_to_joules  #J
        sigma_true = r0   #meter
        alphar0 = alpha*r0
        force = force_Morse
        potential_energy = potential_energy_Morse
    if Potential_formula == 'LJ':
        epsilon_true = 1.65e-21  #1.977e-21 #J
        sigma_true = 3.4e-10 #3.348e-10    #m
        mass = 6.6335209e-26  # kg
        tau = (1.0)/np.sqrt(epsilon_true/(mass*sigma_true*sigma_true)) # time unit in second
        dt = 0.001 * 1.0/tau * 1e-12
        print("time unit is= {0} Seconds\n and timestep={1} femtoseconds".format(tau,tau*dt*1e15))
        force = force_LJ
        potential_energy = potential_energy_LJ
    return None


@njit(parallel=False)
def pbc(X):
    global box_sizes
    X[:,0] -= box_sizes[0]*np.floor(X[:,0]/box_sizes[0])
    X[:,1] -= box_sizes[1]*np.floor(X[:,1]/box_sizes[1])
    X[:,2] -= box_sizes[2]*np.floor(X[:,2]/box_sizes[2])
    return X

@njit(parallel=True)
def force_LJ(X,F):
    F[:,:] = 0.0
    # numpy vectorization slows down this loOP, CAREFUL!
    for i in prange(Natoms):
        for j in prange(i+1,Natoms):
            # vectorize this
            delta_x = X[i,0]-X[j,0]
            delta_y = X[i,1]-X[j,1]
            delta_z = X[i,2]-X[j,2]
            delta_x += (-1 * box_sizes[0])*np.trunc(delta_x/box_half_sizes[0])
            delta_y += (-1 * box_sizes[1])*np.trunc(delta_y/box_half_sizes[1])
            delta_z += (-1 * box_sizes[2])*np.trunc(delta_z/box_half_sizes[2])
            r2 = delta_x**2 + delta_y**2 + delta_z**2
            if np.sqrt(r2)<cutoff:
                f0= 48*(r2**-7 - 0.5*r2**-4)
                fx = delta_x * f0
                fy = delta_y * f0
                fz = delta_z * f0
                F[i, 0] += fx
                F[i, 1] += fy
                F[i, 2] += fz
                F[j, 0] += -fx
                F[j, 1] += -fy
                F[j, 2] += -fz
    return F

@njit(parallel=True)
def force_Morse(X, F):
    global Natoms,alphar0, box_sizes,box_half_sizes,cutoff
    F[:, :] = 0.0
    for i in prange(Natoms):
        for j in prange(i + 1, Natoms):
            # vectorize this
            delta_x = X[i, 0] - X[j, 0]
            delta_y = X[i, 1] - X[j, 1]
            delta_z = X[i, 2] - X[j, 2]
            delta_x += (-1 * box_sizes[0]) * np.trunc(delta_x / box_half_sizes[0])
            delta_y += (-1 * box_sizes[1]) * np.trunc(delta_y / box_half_sizes[1])
            delta_z += (-1 * box_sizes[2]) * np.trunc(delta_z / box_half_sizes[2])

            r = np.sqrt(delta_x ** 2 + delta_y ** 2 + delta_z ** 2)
            if r < cutoff:
                f0 = -2*alphar0 * (np.exp(-2*alphar0*(r-1)) - 2*alphar0*np.exp(-alphar0*(r-1))) 
                fx = delta_x * f0
                fy = delta_y * f0
                fz = delta_z * f0
                F[i, 0] += fx
                F[i, 1] += fy
                F[i, 2] += fz

                F[j, 0] += -fx
                F[j, 1] += -fy
                F[j, 2] += -fz
    return F



@njit(parallel=False)
def potential_energy_LJ(X):
    global Natoms,box_sizes,box_half_sizes,cutoff
    E = 0.
    for i in prange(Natoms):
        for j in range(i+1,Natoms):
            delta_x = X[i,0]-X[j,0]
            delta_y = X[i,1]-X[j,1]
            delta_z = X[i,2]-X[j,2]
            delta_x += (-1 * box_sizes[0])*np.trunc(delta_x/box_half_sizes[0])
            delta_y += (-1 * box_sizes[1])*np.trunc(delta_y/box_half_sizes[1])
            delta_z += (-1 * box_sizes[2])*np.trunc(delta_z/box_half_sizes[2])

            r = np.sqrt(delta_x**2 + delta_y**2 + delta_z**2)
            if r<cutoff:
                E += 4 * ( r**-12 - r**-6)
    return E

@njit(parallel=False)
def potential_energy_Morse(X):
    global Natoms,alphar0, box_sizes,box_half_sizes,cutoff
    E = 0.
    for i in prange(Natoms):
        for j in range(i + 1, Natoms):
            delta_x = X[i, 0] - X[j, 0]
            delta_y = X[i, 1] - X[j, 1]
            delta_z = X[i, 2] - X[j, 2]
            delta_x += (-1 * box_sizes[0]) * np.trunc(delta_x / box_half_sizes[0])
            delta_y += (-1 * box_sizes[1]) * np.trunc(delta_y / box_half_sizes[1])
            delta_z += (-1 * box_sizes[2]) * np.trunc(delta_z / box_half_sizes[2])

            r = np.sqrt(delta_x ** 2 + delta_y ** 2 + delta_z ** 2)
            if r < cutoff:
                E += np.exp(-2*alphar0 * (r - 1)) - 2 * np.exp(-alphar0*(r - 1))
    return E

@njit()
def kinetic_energy(V):
    E = .5*np.sum(V**2)
    return E

@njit()
def temperature(V):
    global Natoms
    return kinetic_energy(V) * 2/(3*Natoms)

@njit()
def thermostat_velocity_rescaling(V):
    global epsilon_true,kB_true,temp_ref
    temp_true = epsilon_true/kB_true # converts to K
    temp_now = temperature(V)*temp_true
    lambda_ = np.sqrt(temp_ref/temp_now)
    V *= lambda_
    return V

@njit(parallel=False)
def velocity_verlet(V,X,F_0):
    global dt,force
    X += dt*V + F_0*dt**2/2
    X = pbc(X) # apply periodic boundary conditions
    F_1 = force(X, F_0)
    V +=  (F_1+F_0) * dt / 2 # calculate velocities at halfstep

    return V,X,F_1

def dump_xyz(f,f3,X,step):
    global Natoms,dump_step

    if step%dump_step!=0:
        return None

    atom_types = Natoms*[1]
    xyz = np.array([atom_types,X[:,0],X[:,1],X[:,2]]).T

    f.write(str(Natoms).encode())
    f.write(b"\n atoms\n")
    np.savetxt(f, xyz,fmt=('%i','%.8f','%.8f','%.8f'))

    dump_xyz_unwrapped(f3,X, step)
    return None


def dump_xyz_unwrapped(f3,X,step):
    global Natoms,box_sizes,box_half_sizes,X0
    X_unwrapped = np.copy(X)
    displacement_M = X-X0
    X_unwrapped[:,0] += (-1 * box_sizes[0]) * np.trunc(displacement_M[:,0] / box_half_sizes[0])
    X_unwrapped[:,1] += (-1 * box_sizes[1]) * np.trunc(displacement_M[:,1] / box_half_sizes[1])
    X_unwrapped[:,2] += (-1 * box_sizes[2]) * np.trunc(displacement_M[:,2] / box_half_sizes[2])

    atom_types = Natoms*[1]
    xyz = np.array([atom_types,X_unwrapped[:,0],X_unwrapped[:,1],X_unwrapped[:,2]]).T

    f3.write(str(Natoms).encode())
    f3.write(b"\n atoms\n")
    np.savetxt(f3, xyz,fmt=('%i','%.8f','%.8f','%.8f'))
    return None


def log(f,X,V,step):
    global epsilon_true,kB_true,log_output,log_step
    global potential_energy
    # print(potential_energy(X))
    if step%log_step!=0:
        return None
    temp_true = epsilon_true/kB_true
    E_kin = kinetic_energy(V)*epsilon_true
    E_pot = potential_energy(X)*epsilon_true
    E_tot = E_kin+E_pot
    temp_now = temperature(V)*temp_true
    log_output = np.array([step,E_kin,E_pot,E_tot,temp_now])
    #print(step,temp_now)
    f.write("\t".join([str(a) for a in log_output])+"\n")
    # np.savetxt(f2, log_output,fmt=('%i','%.8f','%.8f','%.8f','%.8f'))
    return None


def create_atoms(n_a, n_b, n_c):
    unitcell = bulk('Ar', 'fcc', np.power(2,(2./3)),  cubic=True) #,orthorhombic=True)  # ===>  1.122 sigma (from geometry)
    # unitcell = bulk('Ar', 'fcc', 5.4, orthorhombic=True)
    # visualize.view(unitcell)
    # print unitcell.get_cell()
    atoms_a = unitcell
    for i in range(n_a - 1):
        atoms_a = stack(atoms_a, unitcell, axis=0)
    atoms_b = atoms_a
    for j in range(n_b - 1):
        atoms_b = stack(atoms_b, atoms_a, axis=1)
    all_atoms = atoms_b
    for k in range(n_c - 1):
        all_atoms = stack(all_atoms, atoms_b, axis=2)
    # print all_atoms.get_cell()
    print("The cell is= ", all_atoms.get_cell())
    # visualize.view(all_atoms)
    return all_atoms

@njit(parallel=True)
def fix_COM_velocity(V):
    global Natoms
    # calculate CM velocity
    # XMassVelocity, YMassVelocity = 0,0
    XMassVelocity = np.sum(V[:, 0])
    YMassVelocity = np.sum(V[:, 1])
    ZMassVelocity = np.sum(V[:, 2])

    # set CM velocity to zero
    V[:, 0] -= XMassVelocity / Natoms
    V[:, 1] -= YMassVelocity / Natoms
    V[:, 2] -= ZMassVelocity / Natoms
    return  V

def initialize_VXF():
    global n_a,n_b,n_c
    global X0, Natoms
    global box_sizes,box_half_sizes
    atoms = create_atoms(n_a, n_b, n_c)
    Natoms = atoms.get_number_of_atoms()
    print('total number of atoms=', Natoms)
    box_sizes = np.array([atoms.get_cell()[0][0],   atoms.get_cell()[1][1],  atoms.get_cell()[2][2] ] )
    box_half_sizes = box_sizes / 2
    # initialize positions, velocities and forces
    X = atoms.get_positions()
    X0 = np.copy(X)

    # X = np.zeros((Natoms,3))
    V = np.zeros((Natoms,3))
    F = np.zeros((Natoms,3))

    V = np.random.randn(np.shape(V)[0],np.shape(V)[1])
    # print(V)
    V = fix_COM_velocity(V)
    V = thermostat_velocity_rescaling(V)

    return V,X,F

def main():
    global trajectory_file,trajectory_file_unwrapped,log_file
    global Nsteps,temp_step,velocity_zeroing_step
    global force
    global initial_parameteres
    initial_parameteres()

    V,X,F = initialize_VXF()

    #clean old files
    try:
        os.remove(trajectory_file)
        os.remove(trajectory_file_unwrapped)
        os.remove(log_file)
    except OSError:
        pass
    f = open(trajectory_file, "ab")
    f2 = open(trajectory_file_unwrapped, "ab")
    f3 = open(log_file, "a")

    dump_xyz(f,f2,X,0)

    F = force(X,F)

    t_start = time.time()

    for step in range(Nsteps):
        # print(V)
        V,X,F = velocity_verlet(V,X,F)
        dump_xyz(f,f2,X,step)
        log(f3,X,V,step)
        if step % velocity_zeroing_step == 0:
            V = fix_COM_velocity(V)
        if step % temp_step == 0:
            V = thermostat_velocity_rescaling(V)
            #print('yes')
    t_end = time.time()
    print("Time taken: {:.2f} seconds\n".format(t_end-t_start))

    # close files
    f.close()
    f2.close()
    f3.close()
    return None

# object oriented approach, unforunately NUMBA doesnt support classes well so no OOP for now
# class MD(object):
#   def __init__(self,Natoms=20,Nsteps=10**4,box_size=8,dt=10**-5,displacement=2.5,cutoff=2.5,dump_step=10**3,log_step=10**2,thermostat_step=100,temp_ref=160,temp_step=100):
#     self.Natoms = Natoms    #Natoms of atoms
#     self.box_size = box_size # unitless 2D box length
#     self.dt = dt # discretization
#   self.Nsteps =  Nsteps
#     self.displacement = displacement  
#     self.cutoff = cutoff   #cut off
#     self.dump_step = dump_step  
#     self.log_step = log_step  
#     self.thermostat_step = thermostat_step
#     self.temp_ref = temp_step
#     self.temp_step = temp_step    





if __name__ == '__main__':
    main()
